[FIX] pos_self_order: handle selected category when there are no top categories

deni-odoo opened the pull request:
Problem: When the available categories for a point of sale contain one parent category and its child categories and the parent category itself does not have any products, the code filters out the top categories as those that have no parent category. If, in such a case, no products belong to the parent category itself, the selected category is undefined which leads to a blank white screen and the javascript error visible on the console.

Purpose: If there are no top categories, the screen should ideally load the current category computed before. This fix leads to the one of the child categories being the selected category and the kiosk screen loads properly

Steps to Reproduce on Runbot:
1. Create a point of sale. Go to Settings, choose no presets and “Kiosk” in the self-ordering option.
2. Create a parent POS category and two child POS categories.
3. Ensure there are products belonging to the two child categories and none to the parent category
4. Open settings for the POS and choose only these three categories in the “Restrict Categories” section.
5. Go to the point of sale and open the kiosk. Click on “Order Now”. The white screen appears and the js traceback error is visible on the console

opw-5058245

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/228651.png)](https://mergebot.odoo.com/odoo/odoo/pull/228651)

deni-odoo commented:
Hi @manv-afk! thanks, that was a good catch. I have fixed that issue too by changing the availableCategories filter
Previously, only the categories that have products were being loaded which lead to an issue whenever a parent category had no products but its children did. I changed the filter to now include those parent categs whose children have available products. This ensures that the issue you mentioned is resolved and we see the parent categories bar properly and on clicking one of them, it's subsequent child categories in the sub category bar
Code change ->
` .filter(
                (c) =>
                    this.productByCategIds[c.id]?.length > 0 **||
                    c.child_ids.some((child) => child.id in this.productByCategIds)**
            )`
 
<img width="1168" height="572" alt="Screenshot from 2025-09-29 15-27-58" src="https://github.com/user-attachments/assets/53b50075-2fe4-49be-831b-0dd4fb986f8a" />
If the parent categ is not included in the restricted categories of the point of sale, it will not be in the available categories and we will just see one category bar listing all categories
<img width="847" height="444" alt="image" src="https://github.com/user-attachments/assets/b066dc42-1ef4-4bda-85de-d3428a8364f1" />
let me know your feedback!
Thanks.

manv-afk commented on addons/pos_self_order/static/src/app/services/self_order_service.js:
```diff
@@ -160,7 +160,11 @@ export class SelfOrder extends Reactive {
         let now = luxon.DateTime.now();
         now = now.hour + now.minute / 60;
         const availableCategories = this.productCategories
-            .filter((c) => this.productByCategIds[c.id]?.length > 0)
+            .filter(
+                (c) =>
+                    this.productByCategIds[c.id]?.length > 0 ||
+                    c.child_ids.some((child) => child.id in this.productByCategIds)
+            )
             .sort((a, b) => a.sequence - b.sequence);
```
```suggestion
        const isKiosk = this.config.self_ordering_mode === "kiosk";
        const availableCategories = this.productCategories
            .filter(
                (c) =>
                    this.productByCategIds[c.id]?.length > 0 ||
                    (isKiosk && c.child_ids.some((child) => child.id in this.productByCategIds))
            )
            .sort((a, b) => a.sequence - b.sequence);
```

deni-odoo replied:
Hi! I have made the changes, Thanks.

manv-afk reviewed (changes requested):
Hello, thanks for the fix, could you apply the small suggested changes? To handle this case only in Kiosk cause we don't handle category in the same way in self order
